### ISSUE TITLE
useEffect cleanup triggering multiple times after updates?

### ISSUE THREAD

**Issue by akbr** [preactjs/prefresh#503]
While trying to troubleshoot [this issue](https://github.com/preactjs/prefresh/issues/475), I noticed that hmr updates seem to be triggering redundant `useEffect` cleanup calls.

```js
import { useState, useEffect } from 'preact/hooks';

export function App() {
  const [count, setCount] = useState(0);
  const inc = () => setCount((x) => x + 1);

  return <InnerApp count={count} inc={inc} />;
}

function InnerApp({ count, inc }) {
  useEffect(() => {
    console.log('Effect run');
    return () => console.log('Effect cleanup');
  });

  return (
    <div>
      <div>Count is: {count}</div>
      <button onClick={inc}>Increment</button>
    </div>
  );
}
```

If I click "Increment" 20 times, and then trigger an HMR update, the `useEffect` cleanup will run 20 (additional) times:

<img width="228" alt="image" src="https://github.com/preactjs/prefresh/assets/804208/85e19934-f384-4873-980a-6b84a027ac8f">

Sandbox: https://stackblitz.com/edit/vitejs-vite-dxlaxr?file=src%2Fapp.jsx&terminal=dev

**Comment by akbr** [comment]
Another variation:

```js
import { useState, useEffect } from 'preact/hooks';

export function App() {
  const [count, setCount] = useState(0);
  const inc = () => setCount((x) => x + 1);

  return <InnerApp count={count} inc={inc} />;
}

function InnerApp({ count, inc }) {
  useEffect(() => {
    console.log('Effect run');
    return () => console.log('Effect cleanup');
  });

  return (
    <div>
      <div>Count is: {count}</div>
      <button onClick={inc}>Increment</button>
    </div>
  );
}
```

If I click "Increment" 20 times, and then trigger an HMR update, the `useEffect` cleanup will run 20 times:

<img width="228" alt="image" src="https://github.com/preactjs/prefresh/assets/804208/85e19934-f384-4873-980a-6b84a027ac8f">

Sandbox: https://stackblitz.com/edit/vitejs-vite-dxlaxr?file=src%2Fapp.jsx&terminal=dev

**Comment by JoviDeCroock** [comment]
Closed by #504 and #505
